core-bundle/src/Menu/FrontendMenuBuilder.php

```diff
@@ -0,0 +1,314 @@
+            if ($hasSubpages) {
+                $this->getMenu($item, (int) $page->id, $level, $host, $options);
+
+                $childRecords = Database::getInstance()->getChildRecords($page->id, 'tl_page');
+
+                if (
+                    !$options['showLevel']
+                    || $options['showLevel'] >= $level
+                    || (!$options['hardLimit'] && (($requestPage->id === $page->id) || \in_array($requestPage->id, $childRecords, false)))
+                ) {
+                    $item->setDisplayChildren(false);
+                }
+            }
+
+            switch ($page->type) {
+                case 'redirect':
+                    $href = $page->url;
+
+                    if (0 === strncasecmp($href, 'mailto:', 7)) {
+                        $href = StringUtil::encodeEmail($href);
+                    }
+
+                    break;
+
+                case 'root':
+                    // Overwrite the alias to link to the empty URL or language URL (see #1641)
+                    $page->alias = 'index';
+                    $href = $page->getFrontendUrl();
+                    break;
+
+                case 'forward':
+                    if ($page->jumpTo) {
+                        $jumpTo = PageModel::findPublishedById($page->jumpTo);
+                    } else {
+                        $jumpTo = PageModel::findFirstPublishedRegularByPid($page->id);
+                    }
+
+                    // Hide the link if the target page is invisible
+                    if (!$jumpTo instanceof PageModel || (!$jumpTo->loadDetails()->isPublic && !$this->tokenChecker->hasBackendUser())) {
+                        $item->setDisplay(false);
+                    }
+
+                    try {
+                        $href = $jumpTo->getFrontendUrl();
+                    } catch (ExceptionInterface $exception) {
+                        $this->logger->log(LogLevel::ERROR, sprintf('Unable to generate URL for page ID %s: %s', $page->id, $exception->getMessage()), ['contao' => new ContaoContext(__METHOD__, ContaoContext::ERROR)]);
+
+                        continue 2;
+                    }
+
+                    break;
+
+                default:
+                    try {
+                        $href = $page->getFrontendUrl();
+                    } catch (ExceptionInterface $exception) {
+                        $this->logger->log(LogLevel::ERROR, sprintf('Unable to generate URL for page ID %s: %s', $page->id, $exception->getMessage()), ['contao' => new ContaoContext(__METHOD__, ContaoContext::ERROR)]);
+
+                        continue 2;
+                    }
+
+                    break;
+            }
+
+            $this->populateMenuItem($item, $requestPage, $page, $href, $options);
+            $root->addChild($item);
+        }
+
+        // TODO this is the same event as for the backend. must introduce new event
+        $menuEvent = new MenuEvent($this->factory, $root);
```

Comment:
Talking from my tests, you cannot have the same event class dispatched at different spots because this means that the "[simpler event dispatching](https://symfony.com/blog/new-in-symfony-4-3-simpler-event-dispatching)" does not work then. 
Is this correct?

Reply:
Within the event listener, you will check for `$tree->getName()`. The back end main menu is called `mainMenu`. Currently you call the front end menu `root`, but I guess it should be something like `frontEndMenu`.

However, I think within the event you should have access to things like the `ModuleModel` instance for example. So may be we need a new event after all. I think it would be better to have a `BackEndMenuEvent` and a `FrontEndMenuEvent` or something.

Reply:
> I think it would be better to have a BackEndMenuEvent and a FrontEndMenuEvent or something.

:+1: for separating these two concerns.

Reply:
Event added in f6e69a855c1f9ba91f09a063d148d242db55ad79.